core/storage/base_model/gae_models.py

```diff
@@ -1496,6 +1501,46 @@ def get_version(
                 return None
             raise e
 
+    @classmethod
+    def get_version_multi(
+        cls: Type[SELF_VERSIONED_MODEL],
+        entity_ids_versions: List[Tuple[str, Optional[int]]]
+    ) -> List[Optional[SELF_VERSIONED_MODEL]]:
+        entity_ids = set([entity_id for (entity_id, _) in entity_ids_versions])
+        current_version_models = cls.get_multi(
+            entity_ids, include_deleted=False)
+        for current_version_model in current_version_models:
+            current_version_model._require_not_marked_deleted()  # pylint: disable=protected-access
+        latest_versions_by_id = {
+            current_version_model.id: current_version_model.version
+            for current_version_model in current_version_models}
+
+        # The assumption is that every instance of cls uses the same
+        # SNAPSHOT_CONTENT_CLASS, so only one fake instance is needed to fetch
+        # that class for a get_multi() operation.
+        temp_model = cls(id='', version_number=0)
+        assert temp_model.SNAPSHOT_CONTENT_CLASS is not None
+        snapshot_ids = [
+            cls.get_snapshot_id(
+                entity_id,
+                version_number
+                if version_number else latest_versions_by_id[entity_id])
```

Comment:
The optionality here feels odd. We can't make the version number a required input?

(I would generally prefer to keep the core data-fetching functionality simple and easy to audit.)

Reply:
This slightly matches other functionality already in VersionedModel (in that get can take a ``None`` version if I recall). Usually 'None' refers to 'latest' as we don't have a better sentinel value to use for that case. I could introduce a separate API for latest, but that gets a bit messy as it would require two calls (and thus two ``get_multi``s) unless we put the onus on the caller to compile the snapshot IDs.

Do you have a better suggestion?

Reply:
Sorry, my question was unclear. I don't really understand how the controller API is being used, so I wanted to check why the app needs both 'latest' and 'per-version'. Is the latter alone not sufficient?

If not then what you have here is probably the best approach but I wanted to double-check what is actually needed (and why).